<a href="https://www.kaggle.com/code/princeiornongu/s-p-analysis-portfolio-optimization?scriptVersionId=148944269" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# OverView

Disclaimer: The content inside this notebook is Strictly for **DEMONSTRATION ONLY** do not use this as a way for financial investment, **ALWAYS DO YOUR DUE DILIGENCE**.

*   Section 1 is EDA
*   Section 2 is Refined Work
*   Section 3 is Portoflio Optimization: You can out your own Ticker/ Symbols for your own **MOCK PORTFOLIOS**

# Section 1

#### Essential Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

import plotly.express as px
from plotly.subplots import make_subplots

#### Load the data

In [2]:
df = pd.read_csv("/kaggle/input/s-and-p-500-companies-with-financial-information/constituents-financials_csv.csv")
df

Symbol                    Name                  Sector   Price  \
0      MMM              3M Company             Industrials  222.89   
1      AOS         A.O. Smith Corp             Industrials   60.24   
2      ABT     Abbott Laboratories             Health Care   56.27   
3     ABBV             AbbVie Inc.             Health Care  108.48   
4      ACN           Accenture plc  Information Technology  150.51   
..     ...                     ...                     ...     ...   
500    XYL              Xylem Inc.             Industrials   70.24   
501    YUM         Yum! Brands Inc  Consumer Discretionary   76.30   
502    ZBH  Zimmer Biomet Holdings             Health Care  115.53   
503   ZION           Zions Bancorp              Financials   50.71   
504    ZTS                  Zoetis             Health Care   71.51   

     Price/Earnings  Dividend Yield  Earnings/Share  52 Week Low  \
0             24.31        2.332862            7.92       259.77   
1             27.76        1.147959            1.70        68.39   
2             22.51        1.908982            0.26        64.60   
3             19.41        2.499560            3.29       125.86   
4             25.47        1.714470            5.44       162.60   
..              ...             ...             ...          ...   
500           30.94        1.170079            1.83        76.81   
501           27.25        1.797080            4.07        86.93   
502           14.32        0.794834            9.01       133.49   
503           17.73        1.480933            2.60        55.61   
504           32.80        0.682372            1.65        80.13   

     52 Week High    Market Cap        EBITDA  Price/Sales  Price/Book  \
0         175.490  138721055226  9.048000e+09     4.390271       11.34   
1          48.925   10783419933  6.010000e+08     3.575483        6.35   
2          42.280  102121042306  5.744000e+09     3.740480        3.19   
3          60.050  181386347059  1.031000e+10     6.291571       26.14   
4         114.820   98765855553  5.643228e+09     2.604117       10.62   
..            ...           ...           ...          ...         ...   
500        46.860   12915021000  7.220000e+08     2.726209        5.31   
501        62.850   27003303098  2.289000e+09     6.313636      212.08   
502       108.170   24454698119  2.007400e+09     3.164895        2.39   
503        38.430   10670678640  0.000000e+00     3.794579        1.42   
504        52.000   35991109776  1.734000e+09     9.280896       18.09   

                                           SEC Filings  
0    http://www.sec.gov/cgi-bin/browse-edgar?action...  
1    http://www.sec.gov/cgi-bin/browse-edgar?action...  
2    http://www.sec.gov/cgi-bin/browse-edgar?action...  
3    http://www.sec.gov/cgi-bin/browse-edgar?action...  
4    http://www.sec.gov/cgi-bin/browse-edgar?action...  
..                                                 ...  
500  http://www.sec.gov/cgi-bin/browse-edgar?action...  
501  http://www.sec.gov/cgi-bin/browse-edgar?action...  
502  http://www.sec.gov/cgi-bin/browse-edgar?action...  
503  http://www.sec.gov/cgi-bin/browse-edgar?action...  
504  http://www.sec.gov/cgi-bin/browse-edgar?action...  

[505 rows x 14 columns]

Symbol: The unique stock symbol or ticker identifier for each S&P 500 company.

Name: The official name or full corporate title of each company.

Sector: The sector to which the company belongs, categorizing it into specific industry groups within the S&P 500.

Price: The current trading price of the company's stock.

Price/Earnings: The price-to-earnings (P/E) ratio, a key valuation metric, indicating the relationship between the stock's price and its earnings per share.

Dividend Yield: The dividend yield, representing the ratio of the annual dividend payment to the stock's current price.

Earnings/Share: The earnings per share (EPS), a measure of a company's profitability, calculated as earnings divided by the number of outstanding shares.

52 Week Low: The lowest price at which the stock has traded over the past 52 weeks.

52 Week High: The highest price at which the stock has traded over the past 52 weeks.

Market Cap: The total market capitalization of the company, representing the product of the stock's current price and the total number of outstanding shares.

EBITDA: Earnings before interest, taxes, depreciation, and amortization, a measure of a company's operating performance.

Price/Sales: The price-to-sales ratio, which compares the stock's price to the company's revenue per share.

Price/Book: The price-to-book (P/B) ratio, comparing the stock's price to its book value per share, an indicator of the stock's relative value.

#### Explore the Data Part 1

In [3]:
df.describe()

Price  Price/Earnings  Dividend Yield  Earnings/Share  \
count   505.000000      503.000000      505.000000      505.000000   
mean    103.830634       24.808390        1.895953        3.753743   
std     134.427636       41.241081        1.537214        5.689036   
min       2.820000     -251.530000        0.000000      -28.010000   
25%      46.250000       15.350000        0.794834        1.490000   
50%      73.920000       19.450000        1.769255        2.890000   
75%     116.540000       25.750000        2.781114        5.140000   
max    1806.060000      520.150000       12.661196       44.090000   

       52 Week Low  52 Week High    Market Cap        EBITDA  Price/Sales  \
count   505.000000    505.000000  5.050000e+02  5.050000e+02   505.000000   
mean    122.623832     83.536616  4.923944e+10  3.590328e+09     3.941705   
std     155.362140    105.725473  9.005017e+10  6.840544e+09     3.460110   
min       6.590000      2.800000  2.626102e+09 -5.067000e+09     0.153186   
25%      56.250000     38.430000  1.273207e+10  7.739320e+08     1.629490   
50%      86.680000     62.850000  2.140095e+10  1.614399e+09     2.896440   
75%     140.130000     96.660000  4.511968e+10  3.692749e+09     4.703842   
max    2067.990000   1589.000000  8.095080e+11  7.938600e+10    20.094294   

        Price/Book  
count   497.000000  
mean     14.453179  
std      89.660508  
min       0.510000  
25%       2.020000  
50%       3.400000  
75%       6.110000  
max    1403.380000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 505 entries, 0 to 504
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Symbol          505 non-null    object 
 1   Name            505 non-null    object 
 2   Sector          505 non-null    object 
 3   Price           505 non-null    float64
 4   Price/Earnings  503 non-null    float64
 5   Dividend Yield  505 non-null    float64
 6   Earnings/Share  505 non-null    float64
 7   52 Week Low     505 non-null    float64
 8   52 Week High    505 non-null    float64
 9   Market Cap      505 non-null    int64  
 10  EBITDA          505 non-null    float64
 11  Price/Sales     505 non-null    float64
 12  Price/Book      497 non-null    float64
 13  SEC Filings     505 non-null    object 
dtypes: float64(9), int64(1), object(4)
memory usage: 55.4+ KB


In [5]:
df.isnull().sum()

Symbol            0
Name              0
Sector            0
Price             0
Price/Earnings    2
Dividend Yield    0
Earnings/Share    0
52 Week Low       0
52 Week High      0
Market Cap        0
EBITDA            0
Price/Sales       0
Price/Book        8
SEC Filings       0
dtype: int64

#### Fill and convert missing data

In [6]:
# Loop through each column and fill missing values with the mean if the column contains numeric data
for column in df.columns:
    if df[column].dtype in ['int64', 'float64']:
        df[column].fillna(df[column].mean(), inplace=True)

In [7]:
#using a for loop convert all floats to ints
for column in df.columns:
    if df[column].dtype == 'int64':
        df[column] = df[column].astype(float)

#### Explore the Data Part 2

In [8]:
df['Name'].unique()

array(['3M Company', 'A.O. Smith Corp', 'Abbott Laboratories',
       'AbbVie Inc.', 'Accenture plc', 'Activision Blizzard',
       'Acuity Brands Inc', 'Adobe Systems Inc', 'Advance Auto Parts',
       'Advanced Micro Devices Inc', 'AES Corp', 'Aetna Inc',
       'Affiliated Managers Group Inc', 'AFLAC Inc',
       'Agilent Technologies Inc', 'Air Products & Chemicals Inc',
       'Akamai Technologies Inc', 'Alaska Air Group Inc',
       'Albemarle Corp', 'Alexandria Real Estate Equities Inc',
       'Alexion Pharmaceuticals', 'Align Technology', 'Allegion',
       'Allergan, Plc', 'Alliance Data Systems', 'Alliant Energy Corp',
       'Allstate Corp', 'Alphabet Inc Class A', 'Alphabet Inc Class C',
       'Altria Group Inc', 'Amazon.com Inc', 'Ameren Corp',
       'American Airlines Group', 'American Electric Power',
       'American Express Co', 'American International Group, Inc.',
       'American Tower Corp A', 'American Water Works Company Inc',
       'Ameriprise Financial', 'A

In [9]:
df['Sector'].unique()

array(['Industrials', 'Health Care', 'Information Technology',
       'Consumer Discretionary', 'Utilities', 'Financials', 'Materials',
       'Real Estate', 'Consumer Staples', 'Energy',
       'Telecommunication Services'], dtype=object)

In [10]:
 df.groupby('Sector')['Price/Earnings'].mean()
 #code that'll allow grouping S&P 500 companies by their industry and p/e ratio. we see here that Energy Sector is overvalued while the Telecommunionication Sector is undervalued.
 #note that this doesnt give ff the full picture but an average one

Sector
Consumer Discretionary        25.527362
Consumer Staples              21.963235
Energy                        38.559063
Financials                    17.198971
Health Care                   23.460301
Industrials                   21.494328
Information Technology        37.179857
Materials                     20.542400
Real Estate                   22.459091
Telecommunication Services    11.190000
Utilities                     16.846429
Name: Price/Earnings, dtype: float64

In [11]:
 df.groupby('Sector')['Price/Earnings'].mean()

Sector
Consumer Discretionary        25.527362
Consumer Staples              21.963235
Energy                        38.559063
Financials                    17.198971
Health Care                   23.460301
Industrials                   21.494328
Information Technology        37.179857
Materials                     20.542400
Real Estate                   22.459091
Telecommunication Services    11.190000
Utilities                     16.846429
Name: Price/Earnings, dtype: float64

In [12]:
 df.groupby('Sector')['Market Cap'].mean()
 # here we see that Telecommunication Services has the largest market cap with 151,014,200,000 or 151.0142 billion. followed closely by Information Technologies

Sector
Consumer Discretionary        3.825670e+10
Consumer Staples              6.138460e+10
Energy                        4.241605e+10
Financials                    5.062720e+10
Health Care                   5.318621e+10
Industrials                   3.599315e+10
Information Technology        9.610174e+10
Materials                     2.769201e+10
Real Estate                   1.894896e+10
Telecommunication Services    1.510142e+11
Utilities                     2.184402e+10
Name: Market Cap, dtype: float64

In [13]:
 df.groupby('Sector')['Dividend Yield'].mean()
 # we see here that on average Real Estate has the best Yield followed by Utilities

Sector
Consumer Discretionary        1.572412
Consumer Staples              2.433391
Energy                        2.014452
Financials                    2.017241
Health Care                   0.917243
Industrials                   1.479392
Information Technology        1.228494
Materials                     1.737976
Real Estate                   3.894758
Telecommunication Services    7.567797
Utilities                     3.759224
Name: Dividend Yield, dtype: float64

In [14]:
 df.groupby('Sector')['Earnings/Share'].mean()
 # Industrials have the best profitability on a per-share basis on average per sector

Sector
Consumer Discretionary        4.879048
Consumer Staples              3.659412
Energy                       -1.648438
Financials                    4.436912
Health Care                   4.340492
Industrials                   5.344030
Information Technology        3.110000
Materials                     4.932000
Real Estate                   2.391212
Telecommunication Services    4.426667
Utilities                     2.014643
Name: Earnings/Share, dtype: float64

In [15]:
 df.groupby('Sector')['Price/Sales'].mean()
 #The P/S ratio compares a company's market capitalization to its total sales revenue. It can help evaluate whether a stock is overpriced or underpriced relative to its revenue.

Sector
Consumer Discretionary        2.275769
Consumer Staples              2.869688
Energy                        3.689595
Financials                    3.358182
Health Care                   4.827239
Industrials                   2.573401
Information Technology        5.880142
Materials                     2.876093
Real Estate                   9.962681
Telecommunication Services    1.511056
Utilities                     2.561081
Name: Price/Sales, dtype: float64

In [16]:
 df.groupby('Sector')['Price/Book'].mean()
 #A low P/B ratio may indicate potential value, while a high ratio may signal overvaluation. So Utilities on average is a good play while consumer staples would be a bad play on average

Sector
Consumer Discretionary        29.227538
Consumer Staples              56.791564
Energy                         3.190824
Financials                     6.735294
Health Care                    9.094970
Industrials                    9.682978
Information Technology        10.873571
Materials                      4.788800
Real Estate                    6.004848
Telecommunication Services     3.716667
Utilities                      2.106786
Name: Price/Book, dtype: float64

In [17]:
 df.groupby('Sector')['EBITDA'].mean()
 # On average Information Technology comes out on top with a EBITDA of 6,161,996,000 or 6.162 billion.

Sector
Consumer Discretionary        3.501797e+09
Consumer Staples              5.291986e+09
Energy                        4.438905e+09
Financials                    5.290676e+08
Health Care                   3.657468e+09
Industrials                   3.063030e+09
Information Technology        6.161996e+09
Materials                     2.284348e+09
Real Estate                   1.384265e+09
Telecommunication Services    3.365833e+10
Utilities                     3.485105e+09
Name: EBITDA, dtype: float64

In [18]:
df.sort_values(by=['Sector', 'Market Cap'], ascending=[True, False]).head(5)
# here we see the 5 highest market cap companies and they all belong to the Consumer Discretionary Sector and all have market caps larger that the sector on average.

Symbol                     Name                  Sector    Price  \
30    AMZN           Amazon.com Inc  Consumer Discretionary  1350.50   
233     HD               Home Depot  Consumer Discretionary   181.22   
122  CMCSA            Comcast Corp.  Consumer Discretionary    38.19   
442    DIS  The Walt Disney Company  Consumer Discretionary   101.35   
301    MCD         McDonald's Corp.  Consumer Discretionary   158.97   

     Price/Earnings  Dividend Yield  Earnings/Share  52 Week Low  \
30           296.16        0.000000            6.16     1498.000   
233           28.10        1.861049            6.46      207.605   
122           18.54        1.892902            4.74       44.000   
442           17.78        1.603665            5.70      116.100   
301           23.83        2.437994            6.36      178.700   

     52 Week High    Market Cap        EBITDA  Price/Sales  Price/Book  \
30         812.50  6.858734e+11  1.613200e+10     3.927053       24.28   
233        136.82  2.233786e+11  1.651300e+10     2.896440       84.08   
122         34.78  1.864770e+11  2.867500e+10     2.179775        2.65   
442         96.20  1.578173e+11  1.254400e+10     3.963556        3.82   
301        124.36  1.321016e+11  1.051540e+10     5.770237      146.07   

                                           SEC Filings  
30   http://www.sec.gov/cgi-bin/browse-edgar?action...  
233  http://www.sec.gov/cgi-bin/browse-edgar?action...  
122  http://www.sec.gov/cgi-bin/browse-edgar?action...  
442  http://www.sec.gov/cgi-bin/browse-edgar?action...  
301  http://www.sec.gov/cgi-bin/browse-edgar?action...

In [19]:
df['Sector'].value_counts()
# how many companies are in what Sector

Sector
Consumer Discretionary        84
Information Technology        70
Financials                    68
Industrials                   67
Health Care                   61
Consumer Staples              34
Real Estate                   33
Energy                        32
Utilities                     28
Materials                     25
Telecommunication Services     3
Name: count, dtype: int64

#### Visualise the Data

* visual wise putting all S&P 500 companies to be visualised isn't good especially when try to do subplots so a new dataframe will be created only for the top 10 largest companies in the S&P500 which according to +500 are:

  1. Apple Inc. (AAPL)
  2. Microsoft Corp. (MSFT)
  3. Amazon.com Inc. (AMZN)
  4. Meta Inc. (FB)
  5. Tesla Inc.(TSLA)
  6. Alphabet Inc. Class A Shares (GOOGL)
  7. Alphabet Inc. Class C Shares (GOOG)
  8. Berkshire Hathaway Inc. (BRK.B)
  9. Johnson & Johnson (JNJ)
  10. JPMorgan Chase & Co. (JPM)

(https://www.plus500.com/en/Instruments/ES/What-Causes-the-S&P-500%20to-Move~2#:~:text=Some%20things%20to%20consider%20when,political%20events%2C%20and%20interest%20rates.)





In [20]:
fig1 = px.bar(df, x = "Name", y = "Market Cap", color='Name',
             width=1500,height=900, title=" Name vs Market Cap $ B ", )
fig1.show()
# here we can see that some companies are head & shoulders above the rest in terms of Market Cap

In [21]:
fig2 = px.bar(df, x = "Name", y = "EBITDA", color='Name',
             width=1500,height=900, title=" Name vs EDITDA $ B ", )
fig2.show()
# here we see that there are two companies Allergan Plc & FirstEnergy Corp that have negative EBITDA's

In [22]:
fig3 = px.bar(df, x = "Name", y = "Price/Earnings", color='Name',
             width=1500,height=900, title=" Name vs Price/Earnings ", )
fig3.show()
# here we see that there are ten companies that have negative P/E Ratios

In [23]:
fig4 = px.bar(df, x = "Name", y = "Dividend Yield", color='Sector',
             width=1500,height=900, title=" Name vs Dividened Yield by Sector ", )
fig4.show()
# in terms of dividend uield Consumer Discretionary, Utilities & Real Estate are the most consistent

In [24]:
fig5 = px.bar(df, x = "Name", y = "Price/Sales", color='Sector',
             width=1500,height=900, title=" Name vs Price/Sales Grouped by Sector ", )
fig5.show()

#Price/Book

In [25]:
fig6 = px.bar(df, x = "Name", y = "Price/Book", color='Sector',
             width=1500,height=900, title=" Name vs Price/Sales Grouped by Sector ", )
fig6.show()

In [26]:
fig7 = px.bar(df, x = "Sector", y = "Price", color='Name',
             width=1500,height=900, title="Sector vs Price ", )
fig7.show()

In [27]:
fig8 = px.bar(df,x='Name', y='52 Week Low', color='Sector',
              width=1500,height=900, title="Name vs 52 Week Low Grouped by Sector")
fig8.show()

In [28]:
fig10 = px.bar(df,x='Name', y='52 Week High', color='Sector',
              width=1500,height=900, title="Name vs 52 Week Low")
fig10.show()

In [29]:
fig9 = px.pie(df, values='Market Cap', names='Sector', title='Market Cap on Sector with Value')
fig9.show()

In [30]:
fig9 = px.pie(df, values='EBITDA', names='Sector', title='EBIDTA on Sector with Value')
fig9.show()

# Section 2

#### Working with the The top 10 in the S&P500

In [31]:
data = df.sort_values(by='Market Cap', ascending=False)
top10 = data.head(10)
# top10.loc[data['Symbol'] == 'JPM', 'EBITDA'] = 51400000000  # this code here always to input the new EBIDTA. sourced from
top10
# no need to create a new csv file like previous, this is still a good method to use

Symbol                  Name                  Sector    Price  \
51    AAPL            Apple Inc.  Information Technology   155.15   
27   GOOGL  Alphabet Inc Class A  Information Technology  1007.71   
28    GOOG  Alphabet Inc Class C  Information Technology  1001.52   
311   MSFT       Microsoft Corp.  Information Technology    85.01   
30    AMZN        Amazon.com Inc  Consumer Discretionary  1350.50   
187     FB        Facebook, Inc.  Information Technology   171.58   
263    JPM  JPMorgan Chase & Co.              Financials   107.88   
261    JNJ     Johnson & Johnson             Health Care   126.36   
185    XOM     Exxon Mobil Corp.                  Energy    76.07   
66     BAC  Bank of America Corp              Financials    29.74   

     Price/Earnings  Dividend Yield  Earnings/Share  52 Week Low  \
51            16.86        1.579541            9.20     180.1000   
27            31.48        0.000000           22.27    1198.0000   
28            40.29        0.000000           22.27    1186.8900   
311           25.76        1.874791            2.97      96.0700   
30           296.16        0.000000            6.16    1498.0000   
187           27.90        0.000000            5.39     195.3200   
263           15.43        1.984584            6.30     117.3529   
261           17.31        2.556688            0.39     148.3200   
185           21.37        4.003120            1.88      89.3000   
66            16.34        1.536000            1.55      32.6700   

     52 Week High    Market Cap        EBITDA  Price/Sales  Price/Book  \
51       131.1200  8.095080e+11  7.938600e+10     3.458609        5.66   
27       824.3000  7.338240e+11  3.421700e+10     6.801692        4.70   
28       803.1903  7.285356e+11  3.271400e+10     6.772653        4.67   
311       63.2200  6.899784e+11  4.107900e+10     7.113097        9.49   
30       812.5000  6.858734e+11  1.613200e+10     3.927053       24.28   
187      132.4400  5.234230e+11  2.362400e+10    13.241967        7.08   
263       81.6350  3.866136e+11  0.000000e+00     3.371499        1.60   
261      113.1500  3.530625e+11  2.243000e+10     4.632622        4.74   
185       76.0500  3.261487e+11  3.905200e+10     1.770194        1.85   
66        22.0700  3.214782e+11  0.000000e+00     3.201187        1.24   

                                           SEC Filings  
51   http://www.sec.gov/cgi-bin/browse-edgar?action...  
27   http://www.sec.gov/cgi-bin/browse-edgar?action...  
28   http://www.sec.gov/cgi-bin/browse-edgar?action...  
311  http://www.sec.gov/cgi-bin/browse-edgar?action...  
30   http://www.sec.gov/cgi-bin/browse-edgar?action...  
187  http://www.sec.gov/cgi-bin/browse-edgar?action...  
263  http://www.sec.gov/cgi-bin/browse-edgar?action...  
261  http://www.sec.gov/cgi-bin/browse-edgar?action...  
185  http://www.sec.gov/cgi-bin/browse-edgar?action...  
66   http://www.sec.gov/cgi-bin/browse-edgar?action...

##### Making subplot

In [32]:
import plotly.graph_objects as go
# Create subplots
fig = make_subplots(rows=3, cols=4, subplot_titles=top10.columns[3:12], shared_xaxes=False)

# Add traces for financial metrics
for i, column in enumerate(df.columns[3:12]):
    trace = go.Scatter(x=top10['Symbol'], y=top10[column], mode='markers', name=column)
    fig.add_trace(trace, row=i // 4 + 1, col=i % 4 + 1)


fig.update_layout(
    title_text="Financial Metrics Subplots of Top 10 Companies on the S&P500",
    height=900,  # Adjust the overall height of the figure
    width=1500  # Adjust the overall width of the figure
    )
fig.update_xaxes(title_text="Symbol", row=3, col=1)

# Show the subplots
fig.show()
# this code is very import it'll work with whatever data you have and change the values accordingly

##### Creating log returns Chart for the Top 10 using Yahoo Finance



*   Here is showing the movement of Price between the Top 10 from one point in time to another. It'll also be presented in numerical value



In [33]:
!pip install yfinance
# Importing Yahoo Finance
import yfinance as yf

symbols = ['AAPL', 'GOOGL', 'GOOG', 'MSFT', 'AMZN', 'META', 'JPM', 'JNJ', 'XOM', 'BAC'] #FB will be change to META

# Create an empty DataFrame to store the log returns
log_returns = pd.DataFrame()

# Loop through the symbols and fetch data
for symbol in symbols:
    try:
        stock = yf.Ticker(symbol)
        historical_data = stock.history(period="1y")
        if not historical_data.empty:
            close_prices = historical_data['Close']
            log_returns[symbol] = np.log(close_prices / close_prices.shift(1))
        else:
            print(f"No data available for {symbol}")
    except Exception as e:
        print(f"Error fetching data for {symbol}: {str(e)}")

# Create subplots for log returns using Plotly Express
fig = px.line(log_returns, x=log_returns.index, y=log_returns.columns, labels={'index': 'Date', 'value': 'Log Returns'},
              title='Log Returns of Top 10 S&P 500 Companies')

# Show the subplots
fig.show()

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.6/65.6 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 32.1 MB/s eta 0:00:00
  Installing build dependencies ... - \ | / - done
  Getting requirements to build wheel ... - done
  Preparing metadata (pyproject.toml) ... - done
  Created wheel for peewee: filename=peewee-3.17.0-cp310-cp310-linux_x86_64.whl size=290462 sha256=ded66e95a3fb13719899908e4cb665a46dae11c18cd95f870a450e5f5b630b97
  Stored in directory: /root/.cache/pip/wheels/c7/70/ad/212867e96e7004265a69c4aa5dcff00a95f547a67ba26e7e76
Successfully built peewee


##### Calculating the Annual Log Returns and adding it back into the Top 10 S&P500 Dataset

In [34]:
# empty dictionary to store annual log returns
annual_log_returns = {}

# Looping through the symbols and fetch data for each company
for symbol in symbols:
    try:
        stock = yf.Ticker(symbol)
        historical_data = stock.history(period="1y")
        if not historical_data.empty:
            close_prices = historical_data['Close']
            log_returns = np.log(close_prices / close_prices.shift(1))

            # Calculate the annual log return and convert it to a percentage
            log_return_a = log_returns.mean() * 250 * 100

            # Store the result in the dictionary
            annual_log_returns[symbol] = log_return_a

            print(f"Annual Log Return for {symbol}: {log_return_a:.2f}%")
        else:
            print(f"No data available for {symbol}")
    except Exception as e:
        print(f"Error fetching data for {symbol}: {str(e)}")



Annual Log Return for AAPL: 18.79%
Annual Log Return for GOOGL: 37.43%
Annual Log Return for GOOG: 38.20%
Annual Log Return for MSFT: 46.22%
Annual Log Return for AMZN: 39.69%
Annual Log Return for META: 123.67%
Annual Log Return for JPM: 11.95%
Annual Log Return for JNJ: -10.81%
Annual Log Return for XOM: -0.44%
Annual Log Return for BAC: -28.41%


##### adding this back to Top 10


In [35]:
annual_log_returns

{'AAPL': 18.785702777534986,
 'GOOGL': 37.42837033442835,
 'GOOG': 38.195284182240115,
 'MSFT': 46.217353330433056,
 'AMZN': 39.688882106037774,
 'META': 123.67305704244241,
 'JPM': 11.946409245854209,
 'JNJ': -10.810954472294467,
 'XOM': -0.4404127777546757,
 'BAC': -28.41108396387203}

In [36]:
s = pd.Series(annual_log_returns, name='Annual Log Return')
df = s.to_frame()

In [37]:
df


Annual Log Return
AAPL           18.785703
GOOGL          37.428370
GOOG           38.195284
MSFT           46.217353
AMZN           39.688882
META          123.673057
JPM            11.946409
JNJ           -10.810954
XOM            -0.440413
BAC           -28.411084

In [38]:
new_data = df.set_index(top10.index)

top10['Annual Log Return'] = new_data

top10

/tmp/ipykernel_20/4261613228.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Symbol                  Name                  Sector    Price  \
51    AAPL            Apple Inc.  Information Technology   155.15   
27   GOOGL  Alphabet Inc Class A  Information Technology  1007.71   
28    GOOG  Alphabet Inc Class C  Information Technology  1001.52   
311   MSFT       Microsoft Corp.  Information Technology    85.01   
30    AMZN        Amazon.com Inc  Consumer Discretionary  1350.50   
187     FB        Facebook, Inc.  Information Technology   171.58   
263    JPM  JPMorgan Chase & Co.              Financials   107.88   
261    JNJ     Johnson & Johnson             Health Care   126.36   
185    XOM     Exxon Mobil Corp.                  Energy    76.07   
66     BAC  Bank of America Corp              Financials    29.74   

     Price/Earnings  Dividend Yield  Earnings/Share  52 Week Low  \
51            16.86        1.579541            9.20     180.1000   
27            31.48        0.000000           22.27    1198.0000   
28            40.29        0.000000           22.27    1186.8900   
311           25.76        1.874791            2.97      96.0700   
30           296.16        0.000000            6.16    1498.0000   
187           27.90        0.000000            5.39     195.3200   
263           15.43        1.984584            6.30     117.3529   
261           17.31        2.556688            0.39     148.3200   
185           21.37        4.003120            1.88      89.3000   
66            16.34        1.536000            1.55      32.6700   

     52 Week High    Market Cap        EBITDA  Price/Sales  Price/Book  \
51       131.1200  8.095080e+11  7.938600e+10     3.458609        5.66   
27       824.3000  7.338240e+11  3.421700e+10     6.801692        4.70   
28       803.1903  7.285356e+11  3.271400e+10     6.772653        4.67   
311       63.2200  6.899784e+11  4.107900e+10     7.113097        9.49   
30       812.5000  6.858734e+11  1.613200e+10     3.927053       24.28   
187      132.4400  5.234230e+11  2.362400e+10    13.241967        7.08   
263       81.6350  3.866136e+11  0.000000e+00     3.371499        1.60   
261      113.1500  3.530625e+11  2.243000e+10     4.632622        4.74   
185       76.0500  3.261487e+11  3.905200e+10     1.770194        1.85   
66        22.0700  3.214782e+11  0.000000e+00     3.201187        1.24   

                                           SEC Filings  Annual Log Return  
51   http://www.sec.gov/cgi-bin/browse-edgar?action...          18.785703  
27   http://www.sec.gov/cgi-bin/browse-edgar?action...          37.428370  
28   http://www.sec.gov/cgi-bin/browse-edgar?action...          38.195284  
311  http://www.sec.gov/cgi-bin/browse-edgar?action...          46.217353  
30   http://www.sec.gov/cgi-bin/browse-edgar?action...          39.688882  
187  http://www.sec.gov/cgi-bin/browse-edgar?action...         123.673057  
263  http://www.sec.gov/cgi-bin/browse-edgar?action...          11.946409  
261  http://www.sec.gov/cgi-bin/browse-edgar?action...         -10.810954  
185  http://www.sec.gov/cgi-bin/browse-edgar?action...          -0.440413  
66   http://www.sec.gov/cgi-bin/browse-edgar?action...         -28.411084

In [39]:
fig20 = px.bar(top10,x='Symbol', y='Annual Log Return', color='Symbol',
              width=1500,height=900, title="Top 10 vs Annual Log Return")
fig20.show()

# Section 3

#### Portfolio Optimization of the top 10

* Visit: https://builtin.com/data-science/portfolio-optimization-python for more on the coding by Sadrach Pierre


##### Installing the right essential libraries

In [40]:
!pip install PyPortfolioOpt
from pypfopt.efficient_frontier import EfficientFrontier
from pypfopt import risk_models
from pypfopt import expected_returns
from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices
from pypfopt import HRPOpt
from pypfopt.efficient_frontier import EfficientCVaR
from pypfopt.expected_returns import mean_historical_return
from pypfopt.risk_models import CovarianceShrinkage

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.9/61.9 kB 1.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 16.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 298.8/298.8 kB 22.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 48.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.7/10.7 MB 63.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 48.9 MB/s eta 0:00:00
(CVXPY) Nov 01 10:40:40 PM: Encountered unexpected exception importing solver GLOP:
RuntimeError('Version of ortools (9.4.1874) is too old. Expected >= 9.5.0.')
(CVXPY) Nov 01 10:40:40 PM: Encountered unexpected exception importing solver PDLP:
RuntimeError('Version of ortools (9.4.1874) is too old. Expected >= 9.5.0.')


##### Creating a for loop to gather historical Max Close $

In [41]:
# Define a list of stock symbols
symbols = ['AAPL', 'GOOGL', 'GOOG', 'MSFT', 'AMZN', 'META', 'JPM', 'JNJ', 'XOM', 'BAC']

# Create an empty DataFrame to store the closing prices
closing_prices = pd.DataFrame()

# Loop through the symbols and fetch the previous close for each
for symbol in symbols:
    try:
        stock = yf.Ticker(symbol)
        historical_data = stock.history(period="Max")
        if not historical_data.empty:
            # Extract the 'Close' price from the historical data and store it in the DataFrame
            closing_prices[symbol] = historical_data['Close']
        else:
            print(f"No data available for {symbol}")
    except Exception as e:
        print(f"Error fetching data for {symbol}: {str(e)}")

# Assuming you have a DataFrame with a datetime index
# Let's call it 'closing_prices' for the sake of example

closing_prices.index = closing_prices.index.date

# Now, the index of the DataFrame will only contain the date portion
#print(closing_prices)


In [42]:
closing_prices.tail()

AAPL       GOOGL        GOOG        MSFT        AMZN  \
2023-10-26  166.889999  122.279999  123.440002  327.890015  119.570000   
2023-10-27  168.220001  122.169998  123.400002  329.809998  127.739998   
2023-10-30  170.289993  124.459999  125.750000  337.309998  132.710007   
2023-10-31  170.770004  124.080002  125.300003  338.109985  133.089996   
2023-11-01  173.970001  126.449997  127.570000  346.070007  137.000000   

                  META         JPM         JNJ         XOM        BAC  
2023-10-26  288.350006  140.759995  149.000000  107.599998  26.120001  
2023-10-27  296.730011  135.690002  145.600006  105.550003  25.170000  
2023-10-30  302.660004  137.419998  147.029999  105.879997  25.690001  
2023-10-31  301.269989  139.059998  148.339996  105.849998  26.340000  
2023-11-01  311.850006  138.940002  148.690002  105.639999  26.400000

##### Calculating the covariance matrix


*   The Sharpe ratio is the ratio between returns and risk. The lower the risk and the higher the returns, the higher the Sharpe ratio.
*   The higher the Sharpe ratio, the better the performance of the portfolio.



In [43]:
u = expected_returns.mean_historical_return(closing_prices)
S = risk_models.sample_cov(closing_prices)

ef = EfficientFrontier(u, S)
weights = ef.max_sharpe() # optimisezes for maximal sharpe ratio
ef.portfolio_performance(verbose=True)

Expected annual return: 21.7%
Annual volatility: 22.7%
Sharpe Ratio: 0.86


(0.2166790924103469, 0.2273782097440018, 0.8649865465639031)

##### An investment amount of $,1000,000:



*   Here you can put your own specific number for investment for your portfolio
*   Here it's saying to invest 2300 shares in Google Class C and 42 shares in Exxon
*   Here we see that our Annual Expected Return is 21.7%



In [44]:
# here using PyPortfolioOpt to tell us how much of each stock should be bought

latest_prices = get_latest_prices(closing_prices)

da = DiscreteAllocation(weights, latest_prices, total_portfolio_value=1000000)

allocation, leftover = da.greedy_portfolio() #making the most optimal decisions
print("Discrete allocation:", allocation)
print("Funds remaining: ${:.2f}".format(leftover))
#using the 1M$ it's suggested we buy the most in Google stocks etc

Discrete allocation: {'GOOG': 2300, 'JNJ': 1938, 'MSFT': 830, 'AMZN': 713, 'META': 93, 'XOM': 42}
Funds remaining: $69.74


##### Hierarchical Risk Parity (HRP)



*   HRP as a technique for portfolio diversification where the assets are divided and weighted according to a hierarchical tree structure.
*   Expected Annual Return of 17.9% lower than the original 21.7%. However Sharpe Ratio has increased
*   HRP is not as sensitive to outliers
*   HRP suggests we put majority of our money into 218 shares into the **Bank of America** and as little as possible in **Big Tech**



In [45]:
returns = closing_prices.pct_change().dropna()
hrp = HRPOpt(returns)
hrp_weights = hrp.optimize()
hrp.portfolio_performance(verbose=True)
print(dict(hrp_weights))

Expected annual return: 17.9%
Annual volatility: 17.3%
Sharpe Ratio: 0.92
{'AAPL': 0.09629618432861979, 'AMZN': 0.07435155569960679, 'BAC': 0.057332620239645034, 'GOOG': 0.04606430553173201, 'GOOGL': 0.046238028392816384, 'JNJ': 0.31416743318858775, 'JPM': 0.09954606575296267, 'META': 0.03343767024746084, 'MSFT': 0.09551557860601873, 'XOM': 0.13705055801254998}


In [46]:
da_hrp = DiscreteAllocation(hrp_weights, latest_prices, total_portfolio_value=100000)

allocation, leftover = da_hrp.greedy_portfolio()
print("Discrete allocation (HRP):", allocation)
print("Funds remaining (HRP): ${:.2f}".format(leftover))

Discrete allocation (HRP): {'JNJ': 211, 'XOM': 129, 'JPM': 72, 'AAPL': 55, 'MSFT': 28, 'AMZN': 54, 'BAC': 218, 'GOOGL': 36, 'GOOG': 36, 'META': 11}
Funds remaining (HRP): $8.59


##### Mean Conditional Value at Risk (mCVAR)

*   mCVAR works by measuring the worst-case scenarios for each asset in the portfolio
*   we should invest heavily in **META with 184 shares** and then **1 share in Bank of America**


In [47]:
S = closing_prices.cov()
ef_cvar = EfficientCVaR(u, S)
cvar_weights = ef_cvar.min_cvar()

cleaned_weights = ef_cvar.clean_weights()
print(dict(cleaned_weights))

{'AAPL': 0.0, 'GOOGL': 0.0, 'GOOG': 0.0, 'MSFT': 0.42474, 'AMZN': 0.0, 'META': 0.57526, 'JPM': 0.0, 'JNJ': 0.0, 'XOM': 0.0, 'BAC': 0.0}


/opt/conda/lib/python3.10/site-packages/cvxpy/reductions/solvers/solving_chain.py:336: FutureWarning:


    Your problem is being solved with the ECOS solver by default. Starting in 
    CVXPY 1.5.0, Clarabel will be used as the default solver instead. To continue 
    using ECOS, specify the ECOS solver explicitly using the ``solver=cp.ECOS`` 
    argument to the ``problem.solve`` method.
    



In [48]:
da_cvar = DiscreteAllocation(cvar_weights, latest_prices, total_portfolio_value=100000)

allocation, leftover = da_cvar.greedy_portfolio()
print("Discrete allocation (CVAR):", allocation)
print("Funds remaining (CVAR): ${:.2f}".format(leftover))

Discrete allocation (CVAR): {'META': 184, 'MSFT': 123, 'BAC': 1}
Funds remaining (CVAR): $26.59


/opt/conda/lib/python3.10/site-packages/pypfopt/discrete_allocation.py:212: RuntimeWarning:

invalid value encountered in divide

